blockchain/consensus/narwhal_cns/src/proposer.rs

```diff
@@ -0,0 +1,420 @@
+use narwhal_node::{Node, NodeStorage};
+use narwhal_types::{TransactionProto, TransactionsClient};
+
+use crate::consensus::{NarwhalConsensus, NarwhalConsensusError};
+use crate::exec::{ConsensusTransactionIndex, NarwhalExecutionState, NarwhalOutput};
+
+/// `NarwhalProposer` regularly pulls transactions from the mempool, sorted by their
+/// account nonces, and sends them to Narwhal for Atomic Broadcast. The built-in mempool
+/// gossiping is expected to be turned off, so this node will create batches from the
+/// transactions that users sent to it, while other nodes will similarly batch their
+/// own transactions, keeping transaction duplication across batches to the minimum.
+/// Users can send their transactions to multiple nodes, for redundancy and censorship
+/// resistance.
+///
+/// The proposer also acts as the execution engine for the Narwhal library, receiving
+/// committed certificates and batches of transactions from the network. Each certificate
+/// (which contains batches produced by a given validator), will be turned into a single,
+/// deterministic Filecoin block, appended to the end of the chain, then executed.
+///
+/// There might be invalid transactions in the batches, which can be discarded, and
+/// potentially the batching validator punished for them.
+///
+/// Block gossiping is expected to be turned off, so blocks don't need to be signed.
+/// Instead, every validator produces exactly the same blocks, and at some point can
+/// publish their signatures over them, which can be gathered to form a checkpoint.
+pub struct NarwhalProposer {
+    keypair: KeyPair,
+    committee: Committee,
+    validator_addresses: HashMap<PublicKey, Address>,
+    storage_base_path: PathBuf,
+    parameters: Parameters,
+}
+
+#[async_trait]
+impl Proposer for NarwhalProposer {
+    /// Start a Narwhal Primary and a Worker in the background and subscribe to
+    /// local chain extensions. Whenever there is a new tip: (1) check for new
+    /// transactions that can be sent to Narwhal and (2) take the next output
+    /// from Narwhal and turn it into the next block.
+    async fn spawn<DB, MP>(
+        self,
+        state_manager: Arc<StateManager<DB>>,
+        mpool: Arc<MP>,
+        submitter: SyncGossipSubmitter,
+    ) -> anyhow::Result<Vec<JoinHandle<()>>>
+    where
+        DB: BlockStore + Sync + Send + 'static,
+        MP: MessagePoolApi + Send + Sync + 'static,
+    {
+        let chain_store = state_manager.chain_store();
+        let (output_tx, output_rx) = channel::bounded(self.committee.size());
+        let (head_tx, head_rx) = tokio::sync::watch::channel(None);
+        let execution_state = Arc::new(NarwhalExecutionState::new(chain_store.clone(), output_tx));
+
+        // The following is based on the `NodeRestarter` in the narwhal repo.
+
+        // Get a store for the epoch of the committee (although currently there won't be reconfiguration.)
```

Comment:
Once we hook up this to an IPC subnet, we will be able to check the committee from the on-chain state (although we'll have to figure out how to do reconfiguration).

Reply:
They have a `reconfigure.rs` and `restarter.rs` file to showcase. The recommended approach is to stop the current committee and restart a new one.